### PR TITLE
podmanv2 export

### PR DESCRIPTION
**Pull request by baude**
add ability to export a container to a tarball

Signed-off-by: Brent Baude <bbaude@redhat.com>

### CODE REVIEW

File: pkg/api/server/register_containers.go [hunk 1]
```diff
@@ -587,6 +587,29 @@ func (s *APIServer) registerContainersHandlers(r *mux.Router) error {
...
+	//  - in: path
+	//    name: name
+	//    type: string
+	//    required: true
+	//    description: the name or ID of the container
+	// produces:
+	// - application/json
+	// responses:
+	//   200:
+	//     description: ok
```
Comment (jwhonce) [inline pkg/api/server/register_containers.go:606]: The body should contain the tarball.

File: pkg/api/server/register_containers.go [hunk 2]
```diff
@@ -1237,5 +1260,27 @@ func (s *APIServer) registerContainersHandlers(r *mux.Router) error {
...
```
Comment (jwhonce) [inline pkg/api/server/register_containers.go:1279]: Should return the tarball

File: cmd/podmanV2/containers/export.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	if len(exportOpts.Output) == 0 {
+		file := os.Stdout
+		if terminal.IsTerminal(int(file.Fd())) {
+			return errors.Errorf("refusing to export to terminal. Use -o flag or redirect")
+		}
+		exportOpts.Output = "/dev/stdout"
+	} else if err := parse.ValidateFileName(exportOpts.Output); err != nil {
+		return err
+	}
+	return registry.ContainerEngine().ContainerExport(context.Background(), args[0], exportOpts)
```
Comment (jwhonce) [inline cmd/podmanV2/containers/export.go:56]: Why ship the filename upstream? A writer would fit the interface better.  ContainerExport() could just use the writer.

File: pkg/api/handlers/compat/containers_export.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+		utils.Error(w, "failed to save the image", http.StatusInternalServerError, errors.Wrap(err, "failed to save image"))
+		return
+	}
+	rdr, err := os.Open(tmpfile.Name())
+	if err != nil {
+		utils.Error(w, "failed to read temp tarball", http.StatusInternalServerError, errors.Wrap(err, "failed to read the exported tarfile"))
+		return
+	}
+	defer rdr.Close()
+	defer os.Remove(tmpfile.Name())
```
Comment (mheon) [inline pkg/api/handlers/compat/containers_export.go:40]: This should be moved up to right after the creation of the temporary file

File: pkg/bindings/containers/containers.go [hunk 5]
```diff
@@ -296,3 +297,20 @@ func Stop(ctx context.Context, nameOrID string, timeout *uint) error {
 	}
 	return response.Process(nil)
 }
+
+func Export(ctx context.Context, nameOrID string, w io.Writer) error {
```
Comment (mheon) [inline pkg/bindings/containers/containers.go:301]: Needs doc comment

File: pkg/domain/infra/tunnel/containers.go [hunk 6]
```diff
@@ -210,3 +212,18 @@ func (ic *ContainerEngine) ContainerCommit(ctx context.Context, nameOrId string,
...
+	var (
+		err error
+		w   io.Writer
+	)
+	if len(options.Output) > 0 {
+		w, err = os.Create(options.Output)
+		if err != nil {
+			return err
+		}
+
```
Comment (mheon) [inline pkg/domain/infra/tunnel/containers.go:226]: If you spin another version of this, can you get rid of the extra whitespace here?

File: cmd/podmanV2/containers/export.go (hunk 3 above)
Comment (baude) [reply cmd/podmanV2/containers/export.go:56]: while i agree that would be cleaner, the libpod method call does not accept that right now.

**Pull request merged**